### PR TITLE
Created result and subject module at backend...

### PR DESCRIPTION
**Pull request by atharateeq-mergestack**
Create structured modules for "Result" and "Subject" in your application. Define MongoDB schemas using Mongoose, ensuring validations are in place.

**Implement an MVC architecture where:**

- Routes connect to controllers responsible for each module. 
- Each controller should utilize dedicated services to interact with the database efficiently, managing CRUD operations and other data tasks effectively.

### CODE REVIEW

File: src/controllers/resultController.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+export const createResult = async (req: AuthRequest, res: Response) => {
+  try {
+    const resultData = { ...req.body, createdBy: req.userId };
+    const result: IResult | null = await resultService.createResult(resultData);
+    sendResponse(res, HTTP_STATUS.CREATED, MESSAGES.RECORD_ADDED_SUCCESSFULLY, result);
+  } catch (error) {
+    if (error instanceof mongoose.Error.ValidationError) {
+      const errors = Object.values(error.errors).map((err) => err.message);
+      return sendResponse(res, HTTP_STATUS.BAD_REQUEST, errors);
```
Comment (JavedJutt) [inline src/controllers/resultController.ts]: could you change this function to accept parameters as object?

File: src/controllers/resultController.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+export const updateResult = async (req: Request, res: Response) => {
+  try {
+    const resultData = req.body;
+    const resultId = req.params.resultId;
+    const result = await resultService.updateResult(resultId, resultData);
+    if (result) {
+      sendResponse(res, HTTP_STATUS.OK, MESSAGES.RECORD_UPDATED_SUCCESSFULLY, result);
+    } else {
+      sendResponse(res, HTTP_STATUS.NOT_FOUND, MESSAGES.NO_RECORD);
+    }
```
Comment (JavedJutt) [inline src/controllers/resultController.ts]: pass them as object and assign values based on ternary operator

File: src/controllers/resultController.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+export const deleteResult = async (req: Request, res: Response) => {
+  try {
+    const resultId = req.params.resultId;
+    await resultService.deleteResult(resultId);
+    sendResponse(res, HTTP_STATUS.OK, MESSAGES.RECORD_DELETED_SUCCESSFULLY, {});
+  } catch (error) {
+    sendResponse(res, HTTP_STATUS.INTERNAL_SERVER_ERROR, MESSAGES.INTERNAL_SERVER_ERROR);
+  }
+};
```
Comment (JavedJutt) [inline src/controllers/resultController.ts]: pls review your PR yourself before sending to review again eof.

File: src/controllers/subjectController.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    sendResponse(res, HTTP_STATUS.CREATED, MESSAGES.RECORD_ADDED_SUCCESSFULLY, subject);
+  } catch (error) {
+    if (error instanceof mongoose.Error.ValidationError) {
+      const errors = Object.values(error.errors).map((err) => err.message);
+      return sendResponse(res, HTTP_STATUS.BAD_REQUEST, errors);
+    } else {
+      return sendResponse(res, HTTP_STATUS.INTERNAL_SERVER_ERROR, MESSAGES.INTERNAL_SERVER_ERROR);
+    }
+  }
+};
```
Comment (JavedJutt) [inline src/controllers/subjectController.ts]: can we avoid this if else statement everywhere? by creating `sendErrorResponse` function or by any other approach?

File: src/controllers/subjectController.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      sendResponse(res, HTTP_STATUS.NOT_FOUND, MESSAGES.NO_RECORD);
+    }
+  } catch (error) {
+    if (error instanceof mongoose.Error.ValidationError) {
+      const errors = Object.values(error.errors).map((err) => err.message);
+      return sendResponse(res, HTTP_STATUS.BAD_REQUEST, errors);
+    } else {
+      return sendResponse(res, HTTP_STATUS.INTERNAL_SERVER_ERROR, MESSAGES.INTERNAL_SERVER_ERROR);
+    }
+  }
```
Comment (JavedJutt) [inline src/controllers/subjectController.ts]: ditto

File: src/middleware/authMiddleware.ts [hunk 6]
```diff
@@ -11,6 +13,9 @@ export const authenticateUser = (req: AuthRequest, res: Response, next: NextFunc
     return sendResponse(res, HTTP_STATUS.UNAUTHORIZED, MESSAGES.UNAUTHORIZED);
   }  
   const token = authHeader.split(' ')[1];
```
Comment (JavedJutt) [inline src/middleware/authMiddleware.ts]: splitting and accessing first index in an array can we do it by loadash?

File: src/utils/enum.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+export enum Grade {
+    A_PLUS = 'A+',
+    A_MINUS = 'A-',
+    B_PLUS = 'B+',
+    B_MINUS = 'B-',
+    F = 'F'
+  }
+
```
Comment (JavedJutt) [inline src/utils/enum.ts]: can you check why this sign is appearing here

**Review by JavedJutt — CHANGES_REQUESTED** [review]
avoid if else statements

File: src/utils/enum.ts (hunk 7 above)
Comment (atharateeq-mergestack) [reply src/utils/enum.ts]: In EOF one space is causing this sign to show.

**Review by JavedJutt — APPROVED** [review]

**Pull request merged**